pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+        """This method converts a DmapArray to the byte format to be written
+        out.
+
+        The format is name,then type, number of dimensions, dimensions,
+        array data.
+
+        :param array: a DmapArray
+        :returns: total bytes the array will take up
+
+        """
+
+        name = "{0}\0".format(array.get_name())
+        struct_fmt = '{0}s'.format(len(name))
+        name_bytes = struct.pack(struct_fmt, name.encode('utf-8'))
+
+        dmap_type_bytes = struct.pack('c',
+                                      chr(array.get_type()).encode('utf-8'))
+
+        dimension_bytes = struct.pack('i', array.get_dimension())
+        arr_dimensions_bytes = bytes()
+        for dim in array.get_arr_dimensions():
+            arr_dimensions_bytes = arr_dimensions_bytes + struct.pack('i', dim)
+
+        data_bytes = array.get_data().tostring()
+
+        total_bytes = name_bytes + dmap_type_bytes + dimension_bytes +\
+            arr_dimensions_bytes + data_bytes
+        return total_bytes
+
+    def type_to_fmt(self, data):
+        """Finds data types and converts them to a format specifier for
+        struct or numpy
+        packing methods
+
+        :param data: data for which to find type
+        :returns: a string format identifier for the python data type
+
+        """
+        if isinstance(data, int):
+            return 'i'
+        elif isinstance(data, str):
+            return 's'
+        elif isinstance(data, float):
+            return 'f'
+        elif isinstance(data, np.float32):
+            return 'f'
+        elif isinstance(data, np.float64):
+            return 'd'
+        elif isinstance(data, np.char):
+            return 'c'
+        elif isinstance(data, np.int8):
+            return 'c'
+        elif isinstance(data, np.int16):
+            return 'h'
+        elif isinstance(data, np.int32):
+            return 'i'
+        elif isinstance(data, np.int64):
+            return 'q'
+        elif isinstance(data, np.uint8):
+            return 'B'
+        elif isinstance(data, np.uint16):
+            return 'H'
+        elif isinstance(data, np.uint32):
+            return 'I'
+        elif isinstance(data, np.uint64):
+            return 'Q'
+        else:
+            return ''
+
+    def convert_fmt_to_dmap_type(self, fmt):
+        """Converts a format specifier to a dmap type to be written
+        as part of buffer
+
+        :param fmt: a string format identifier for the DMAP data type
+        :returns: DMAP type
+
+        """
+        return {
+                'c': CHAR,
+                'h': SHORT,
+                'i': INT,
+                'f': FLOAT,
+                'd': DOUBLE,
+                's': STRING,
+                'q': LONG,
+                'B': UCHAR,
+                'H': USHORT,
+                'I': UINT,
+                'Q': ULONG,
+            }.get(fmt, None)
+
+
+# TODO: consider removing this function and get the user to call RawDmapWrite
```

Comment:
Kinda disagree. Maybe this could be refactored, but this method has all the types and fields for the common SD formats. In theory all fields should be cast to these types when writing, so the user can use numpy or python types in their analysis without having to worry if their output is going to be compatible with RST.

Reply:
Out of scope. We shall see what happens when I clean up DmapWrite.